beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -241,57 +437,56 @@ private void startObserve(String name, Partition partition, OffsetRange restrict
       observers.put(partition.getId(), observer);
       final ObserveHandle handle;
       if (restriction.getStartOffset() != null) {
-        handle =
-            reader.observeBulkOffsets(
-                Collections.singletonList(restriction.getStartOffset()), observer);
+        handle = observeBulkOffsets(restriction, reader, observer);
         partitionToSeekedOffset.put(partition.getId(), restriction.getStartOffset());
       } else {
-        handle =
-            reader.observeBulkPartitions(
-                name,
-                Collections.singletonList(restriction.getPartition()),
-                restriction.getPosition(),
-                observer);
+        handle = observeBulkPartitions(name, restriction, reader, observer);
       }
       runningObserves.put(partition.getId(), handle);
     }
 
-    @GetInitialRestriction
-    public OffsetRange initialRestriction() {
-      return OffsetRange.initialRestriction(limit);
+    abstract ObserveHandle observeBulkOffsets(
+        OffsetRange restriction, CommitLogReader reader, BlockingQueueLogObserver observer);
+
+    abstract ObserveHandle observeBulkPartitions(
+        String name,
+        OffsetRange restriction,
+        CommitLogReader reader,
+        BlockingQueueLogObserver observer);
+
+    public void setup() {
+      runningObserves = new HashMap<>();
+      partitionToSeekedOffset = new HashMap<>();
+      observers = new HashMap<>();
     }
 
-    @SplitRestriction
-    public void splitRestriction(
-        @Restriction OffsetRange restriction, OutputReceiver<OffsetRange> splits) {
+    public void tearDown() {
+      Lists.newArrayList(observers.keySet()).forEach(p -> closeHandle(p, true));
```

Comment:
Set has a `forEach` method as well?

Reply:
Hm, no idea why it previously cloned the key set. I suppose that was part of some debugging.

Reply:
I have idea now. The `closeHandle` removes the observer from the `observers`, so not cloning the key set results in ConcurrentModificationException.